File: projectEuler/14/14.py
#Longest Collatz Sequence
'''
Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.
'''
#(15.313s)
def collatz(n):
    if n % 2 ==0:
        return n/2 #n is even
    else:
        return (3*n)+1 #n is odd

def collatz_sequence(n):
    count = 1
    while n !=1:
        n = collatz(n)
        count += 1
    return count

def longest_collatz_sequence(n):
    max_count = 0
    max_num = 0
    for i in range(1,n):
        count = collatz_sequence(i)
        if count > max_count:
            max_count = count
            max_num = i
    return max_num

print(longest_collatz_sequence(1000000))


# Faster Solution (15.263s)
# def collatz(n):
#     if n % 2 == 0:
#         return n // 2  # n is even
#     else:
#         return 3 * n + 1  # n is odd

# def collatz_sequence_length(n, memo):
#     if n in memo:
#         return memo[n]
    
#     if n == 1:
#         return 1
    
#     next_n = collatz(n)
#     length = 1 + collatz_sequence_length(next_n, memo)
#     memo[n] = length
    
#     return length

# def longest_collatz_sequence(limit):
#     max_count = 0
#     max_num = 0
#     memo = {}
    
#     for i in range(1, limit):
#         count = collatz_sequence_length(i, memo)
#         if count > max_count:
#             max_count = count
#             max_num = i
            
#     return max_num

# print(longest_collatz_sequence(1000000))

File: projectEuler/14/14.cpp
#include <iostream>
#include <unordered_map>

// Function to compute the next number in the Collatz sequence
unsigned long long collatz(unsigned long long n) {
    return (n % 2 == 0) ? (n / 2) : (3 * n + 1);
}

// Function to compute the length of the Collatz sequence using memoization
unsigned long long collatzSequenceLength(unsigned long long n, std::unordered_map<unsigned long long, unsigned long long>& memo) {
    if (memo.find(n) != memo.end()) {
        return memo[n];
    }
    
    if (n == 1) {
        return 1;
    }
    
    unsigned long long next_n = collatz(n);
    unsigned long long length = 1 + collatzSequenceLength(next_n, memo);
    memo[n] = length;
    
    return length;
}

// Function to find the starting number under the given limit that produces the longest Collatz sequence
unsigned long long longestCollatzSequence(unsigned long long limit, std::unordered_map<unsigned long long, unsigned long long>& memo) {
    unsigned long long max_count = 0;
    unsigned long long max_num = 0;
    
    for (unsigned long long i = 1; i < limit; ++i) {
        unsigned long long count = collatzSequenceLength(i, memo);
        if (count > max_count) {
            max_count = count;
            max_num = i;
        }
    }
    
    return max_num;
}

int main() {
    unsigned long long limit = 1000000;
    std::unordered_map<unsigned long long, unsigned long long> memo;

    unsigned long long result = longestCollatzSequence(limit, memo);
    
    // Calculate approximate memory usage
    size_t memory_usage = memo.size() * (sizeof(unsigned long long) + sizeof(unsigned long long) + 16); // Approximate overhead

    std::cout << "The starting number under " << limit << " that produces the longest Collatz sequence is: " << result << std::endl;
    std::cout << "Approximate memory usage: " << memory_usage << " bytes (" << memory_usage / (1024.0 * 1024.0) << " MB)" << std::endl;

    return 0;
}